quinn-proto/src/connection/streams/types.rs

```diff
@@ -0,0 +1,528 @@
+    fn bytes_read(&self) -> u64;
+}
+
+impl BytesRead for usize {
+    fn bytes_read(&self) -> u64 {
+        *self as u64
+    }
+}
+
+impl BytesRead for Bytes {
+    fn bytes_read(&self) -> u64 {
+        self.len() as u64
+    }
+}
+
+impl BytesRead for (Bytes, u64) {
+    fn bytes_read(&self) -> u64 {
+        self.0.len() as u64
+    }
+}
+
+pub(crate) struct ReadChunks {
+    pub bufs: usize,
+    pub read: usize,
+}
+
+impl BytesRead for ReadChunks {
+    fn bytes_read(&self) -> u64 {
+        self.read as u64
+    }
+}
+
+/// Errors triggered when reading from a recv stream
+#[derive(Debug, Error, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum ReadError {
+    /// No more data is currently available on this stream.
+    ///
+    /// If more data on this stream is received from the peer, an `Event::StreamReadable` will be
+    /// generated for this stream, indicating that retrying the read might succeed.
+    #[error("blocked")]
+    Blocked,
+    /// The peer abandoned transmitting data on this stream.
+    ///
+    /// Carries an application-defined error code.
+    #[error("reset by peer: code {}", 0)]
+    Reset(VarInt),
+    /// The stream has not been opened or was already stopped, finished, or reset
+    #[error("unknown stream")]
+    UnknownStream,
+    /// Attempted an ordered read following an unordered read
+    ///
+    /// Performing an unordered read allows discontinuities to arise in the receive buffer of a
+    /// stream which cannot be recovered, making further ordered reads impossible.
+    #[error("ordered read after unordered read")]
+    IllegalOrderedRead,
+}
+
+impl From<IllegalOrderedRead> for ReadError {
+    fn from(_: IllegalOrderedRead) -> Self {
+        ReadError::IllegalOrderedRead
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub(super) enum SendState {
+    /// Sending new data
+    Ready,
+    /// Stream was finished; now sending retransmits only
+    DataSent { finish_acked: bool },
+    /// Sent RESET
+    ResetSent,
+    /// All sent data acknowledged
+    DataRecvd,
+    /// Reset acknowledged
+    ResetRecvd,
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+enum RecvState {
+    Recv { size: Option<u64> },
+    ResetRecvd { size: u64, error_code: VarInt },
+    Closed,
+}
+
+impl Default for RecvState {
+    fn default() -> Self {
+        RecvState::Recv { size: None }
+    }
+}
+
+/// Reasons why attempting to finish a stream might fail
+#[derive(Debug, Error, Clone, PartialEq, Eq)]
+pub enum FinishError {
+    /// The peer is no longer accepting data on this stream. No
+    /// [`StreamEvent::Finished`] event will be emitted for this stream.
+    ///
+    /// Carries an application-defined error code.
+    ///
+    /// [`StreamEvent::Finished`]: crate::StreamEvent::Finished
+    #[error("stopped by peer: code {}", 0)]
+    Stopped(VarInt),
+    /// The stream has not been opened or was already finished or reset
+    #[error("unknown stream")]
+    UnknownStream,
+}
```

Comment:
I'd like to see this broken up even further, into separate send/recv modules. I've spent a lot of time lost/confused due to file-scoped searches turning up states for the wrong type of stream.

Reply:
👍 done.